Fast-path ExecutionContext for ThreadPool items

benaadams opened the pull request:
`ExecutionContext.Run` has logic for saving and restoring the current contexts and extra logic to optimise the enregistering around exception handling.

When items are run on the ThreadPool we can fast-path for the threads starting on the Default context and avoid stashing the current context to restore later (as it is Default). 

For CallbackDefaultContext items we can further fast-path by not restoring the new context (as it is still the Default) and only need to return back to the Default context after the callback is run (in case it moves off the Defaults in the callback).

Additionally; if a naked exception is thrown on the ThreadPool it will crash the process; so the additional exception handling logic can be avoided.

The `ThreadPool.Dispatch` loop resets `ExecutionContext` and `SynchronizationContext` after the workitems to move the ThreadPool thread back onto the Default context from any `AsyncLocal<T>` leakage from Unsafe work items, CustomWorkItems or AsyncLocals being set in AsyncLocals `OnValuesChanged` handler to guarantee the ThreadPool threads are on the Default contexts when a workitem starts. * 

In addition to the standard `ExecutionContext.RunInternal` (328 bytes asm) this introduces 3 fast-path overloads for running via `ExecutionContext` from the ThreadPool using the above preconditions.

`ExecutionContext:RunFromThreadPoolDispatchLoop` (230 bytes asm), for 
`Task`/`AsyncStateMachineBox`/`AwaitTaskContinuation`/`Timer` using the `Task.ExecuteFromThreadPool` overload; where any exception must be captured and propagated after the Default context has been restored.

`ExecutionContext:RunForThreadPoolUserWorkItem` (108 bytes asm), for QueueUserWorkItems that aren't on the default context.

`ExecutionContext:RunForThreadPoolUserWorkItemWithDefaultContext` (71 bytes asm), for QueueUserWorkItems that are on the default context.

Also drops an additional delegate invocation for work items queued using the  `QueueUserWorkItem<TState>(Action<TState>, TState, Boolean)` overload by accepting the `Action<TState>` directly rather than requiring a wrapper to convert it to a `ContextCallback` delegate.

Resolves:  dotnet/corefx#32695

*In `Dispatch` the currentThread is stored and used from the same lookup as the work queue to avoid calling `Thread.CurrentThread`  VM call https://github.com/dotnet/coreclr/pull/20308#issuecomment-435769584; and added some tweaks to the loop to enregister the locals https://github.com/dotnet/coreclr/pull/20308/commits/5db2e9f94ac3d32f1aad9957486b7893dc426bb8 to make up the difference.

/cc @stephentoub @kouvel

stephentoub commented on src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:
```diff
@@ -51,6 +52,37 @@ public void GetObjectData(SerializationInfo info, StreamingContext context)
+        internal static void ResetToDefaults()
+        {
+            Thread currentThread = Thread.CurrentThread;
+            if (currentThread.ExecutionContext != null || currentThread.SynchronizationContext != null)
```
These checks are now being done for all work items, regardless of whether they correctly reset or not?  Assuming everything is done correctly, then, this would be pure overhead in some cases, right?  What's the worst-case throughput impact in that case?

benaadams replied:
The `null` checks should be fairly nominal and they should always be `null` (as `ExecutionContext.Run` will reset it back)?

However, I think `Thread.CurrentThread` isn't a low cost call...

stephentoub replied:
> However, I think Thread.CurrentThread isn't a low cost call...

Right, that's the main thing I was referring to.

stephentoub commented on src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:
```diff
@@ -51,6 +52,37 @@ public void GetObjectData(SerializationInfo info, StreamingContext context)
+        [MethodImpl(MethodImplOptions.AggressiveInlining)]
+        internal static void ResetToDefaults()
+        {
+            Thread currentThread = Thread.CurrentThread;
```
ResetToDefaults is called in a loop.  With this aggressively inlined, does this somehow get hoisted as an invariant out of the loop?  Or if not, should it instead be passed in as an argument and grabbed once, since it's a TLS hit?

benaadams replied:
Good idea

benaadams commented:
Since we have the thread (with this change); could propagate it into some EC calls and make back the difference

benaadams commented:
Ah, no we can't as the workitems are self contained

stephentoub commented:
Does this have any measurable impact on perf? What's the worst case impact?

benaadams commented:
> Does this have any measurable impact on perf? What's the worst case impact?

Not sure yet. 50/50 on whether this is worth perusing anyway; it would be an unusual state to get into.

Do you think the new `TheadPool.UnsafeQueueCustomWorkItem` and `IThreadPoolWorkItem` would increase the chance of knocking the threadpool thread off the Default execution context?

benaadams commented:
Could maybe combine it with a change to QUWI to use an ExecutionContext.Run that always assumes a `null` (Default) context? Might make it more worthwhile.

benaadams commented:
Rather `QueueUserWorkItemCallback` et al; that assume the thread starts on Default context so it doesn't have to save the current state and can use implied nulls for EC.Run initial state

stephentoub commented:
> Do you think the new TheadPool.UnsafeQueueCustomWorkItem and IThreadPoolWorkItem would increase the chance of knocking the threadpool thread off the Default execution context?

I don't think so.  I'd expect anything that's able to use `ThreadPool.UnsafeQueueUserWorkItem(IThreadPoolWorkItem, bool)` was previously using (or able to use)  `ThreadPool.UnsafeQueueUserWorkItem(WaitCallback,object)`, and they both have the same semantics with regards to ExecutionContext.

benaadams commented:
Going a different direction

benaadams commented:
Updated summary for new approach

----
ExecutionContext.Run has logic for saving and restoring the current contexts and extra logic to optimise the enregistering around exception handling.

When items are run on the ThreadPool we can fast-path for the threads starting on the Default context and avoid stashing the current context to restore later (as it is Default). 

For CallbackDefaultContext items we can further fast-path by not restoring the new context (as it is still the Default) and only need to return back to the Default context after the callback is run (in case it moves off the Defaults in the callback).

Additionally; if a naked exception is thrown on the ThreadPool it will crash the process; so the additional exception handling logic can be avoided.

Note: I have added guard logic for the ThreadPool thread not being on the default context as Unsafe queued items can move these threads off the Default if the callback is not constructed correctly. When it moves back it in the guard it will not send change notifications. This is a change to current where it will send a change notification for every thread-pool item (as it would be moving from the modified ThreadPool thread to the requested context e.g. Default)

Unsafe modifications to the ThreadPool threads could still be seen between unsafe executions  dotnet/corefx#32695; however regular executions will eventually reset the ThreadPool thread back to the Defaults.

----
@stephentoub PTAL

benaadams commented:
My new assertion is failing
```
03:51:28 Discovering: System.Memory.Tests
03:51:34 Discovered:  System.Memory.Tests
03:51:39 Assertion Failed
03:51:39 UnsafeQueueUserWorkItem has changed the ExecutionContext and not restored it.
```

benaadams commented:
Looks like might be from `System.Threading.Channels.Tests`
```
06:04:28 Starting:    System.Threading.Channels.Tests
06:04:29 Assertion Failed
06:04:29 UnsafeQueueUserWorkItem has changed the ExecutionContext and not restored it.
06:04:29 
06:04:29    at System.Threading.ExecutionContext.RunFromThreadPool(ExecutionContext executionContext, ContextCallback callback, Object state)
06:04:29    at System.Threading.ThreadPoolWorkQueue.Dispatch()
06:04:32    System.Threading.Channels.Tests.StressTests.RunInStressMode [SKIP]
```

benaadams commented:
Trying to catch where it moves the threadpool to non-default

stephentoub commented:
> Looks like might be from System.Threading.Channels.Tests

There are some tests of ExecutionContext and SynchronizationContext that are lazy about resetting things... maybe one of those? e.g. https://github.com/dotnet/corefx/blob/dc9d6577b3ac8c3cd9e92164b9c7e9200bcbc764/src/System.Threading.Channels/tests/ChannelTestBase.cs#L1080

benaadams commented:
I think there are a few bits
```
07:59:57 Starting:    Canonicalization.Tests
07:59:59 Assertion Failed
07:59:59 
07:59:59    at System.Threading.TimerQueueTimer.CallCallback()
07:59:59    at System.Threading.TimerQueueTimer.Fire()
07:59:59    at System.Threading.TimerQueue.FireNextTimers()
```

benaadams commented:
Yeah looks like it might be channels
```
19:33:49 Discovering: System.Threading.Channels.Tests
19:33:51 Assertion Failed
19:33:51 Flow suppressed Task has changed the ExecutionContext and not restored it.
19:33:51 
19:33:51    at System.Threading.ThreadPoolWorkQueue.Dispatch()
19:33:51 Assertion Failed
19:33:51 Flow suppressed Task has changed the ExecutionContext and not restored it.
```

benaadams commented:
The test `AsyncLocalTests.SetValueFromNotification` is designed to leak the `AsyncLocal<T>` as it changes it in the `valuesChangedHandler` callback; to allow that behaviour a reset would need to occur in the `Dispatch` loop after the work item has completed. Which would also fix https://github.com/dotnet/corefx/issues/32695

stephentoub commented:
> to allow that behaviour a reset would need to occur in the Dispatch loop

To test what behavior?  The test is passing.

benaadams commented:
>  The test is passing.

:) Causes the asserts in this PR to fail as it changes the Threadpool thread from the default context after its been reset to it (it then passes the changed context via the `await`). 

The current behaviour means that the `AsyncLocal<T>` forever captured on that Threadpool thread and every item then run on it will then trigger the `valuesChangedHandler`; which probably isn't desirable anyway. Resetting it in the Dispatch loop after the workitem completes, still allows the test to pass,

benaadams commented:
`ExecutionContext.Run` current asm
```
; Total bytes of code 328, prolog size 28 for method ExecutionContext:RunInternal(ref,ref,ref)
```

New ThreadPool fast paths asm:

Called by `ExecuteFromThreadPool` for `Task`/`AsyncStateMachineBox`/`AwaitTaskContinuation`/`Timer`
```
; Total bytes of code 243, prolog size 26 for method :RunFromThreadPool(ref,ref,ref)
```
QUWI non-default context
```
; Total bytes of code 124, prolog size 8 for method :RunUserWorkItem(ref,ref,byref)
```
QUWI default context
```
; Total bytes of code 71, prolog size 5 for method :RunUserWorkItemWithDefaultContext(ref,byref)
```

benaadams commented:
> > However, I think Thread.CurrentThread isn't a low cost call...

> Right, that's the main thing I was referring to.

Caching it in `ThreadPoolWorkQueueThreadLocals` now

benaadams commented:
Updated summary

kouvel commented on src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:
```diff
@@ -566,6 +570,17 @@ internal static bool Dispatch()
                     //
                     workQueue.EnsureThreadRequested();
 
+                    // Start on clean ExecutionContext and SynchronizationContext
```
Would it be better to clear the contexts after a work item such that they can be collected if there are no more work items and the thread pool goes idle?

benaadams replied:
I originally had that but the Assert failed that the EC wasn't clean; at best I can tell unmanged items also use the threadpool threads and one of the timers from the `AppDomainTimerCallback` doesn't run from the `Dispatch` loop, but uses the thread its on (that does the dispatching to the ThreadPool) to run the last timer; so one of those might have "leaked" a change.

kouvel replied:
Makes sense, but the values can be cleared once before the dispatch loop and thereafter after each work item completes. Don't think it would make a big difference but maybe slightly better.

benaadams replied:
Ah, good idea

kouvel commented on src/System.Private.CoreLib/src/System/Threading/Timer.cs:
```diff
@@ -674,9 +674,9 @@ public ValueTask CloseAsync()
+        void IThreadPoolWorkItem.Execute() => Fire(isThreadPool: true);
 
-        internal void Fire()
+        internal void Fire(bool isThreadPool = false)
```
All calls to this should be from thread pool threads. Could add an assert to `FireNextTimers()` that it's a thread pool thread, and always use `RunFromThreadPool()` in `CallCallback()`.

benaadams replied:
One `Fire` is not called from the `Dispatch` loop

benaadams commented on src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:
```diff
@@ -136,6 +136,11 @@ internal static void RunInternal(ExecutionContext executionContext, ContextCallb
             Thread currentThread0 = Thread.CurrentThread;
             Thread currentThread = currentThread0;
             ExecutionContext previousExecutionCtx0 = currentThread0.ExecutionContext;
+            if (previousExecutionCtx0?.m_isDefault ?? false)
```
This compiled to (very slightly) better asm for some reason

benaadams replied:
Just rechecked, they both generate identical asm in .NET Core 3.0 so I can switch back.

Is a shame the `?? false` bit is needed 😞

benaadams replied:
Changed

benaadams commented on src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:
```diff
@@ -282,6 +292,133 @@ internal static void RunInternal<TState>(ExecutionContext executionContext, Cont
+            }
+
+            ExecutionContext currentExecutionCtx = currentThread1.ExecutionContext;
+            if (currentExecutionCtx != null && !currentExecutionCtx.m_isDefault)
```
resetting when `currentExecutionCtx != null && currentExecutionCtx.m_isDefault` drops the default test so is probably a good idea also (one less branch)

benaadams replied:
Changed both

benaadams commented:
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

stephentoub commented:
Where are we on determining perf impact (positive or negative)? :smile:

benaadams commented on src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:
```diff
@@ -282,6 +292,139 @@ internal static void RunInternal<TState>(ExecutionContext executionContext, Cont
+
+            // Capture current thread to local rather than looking it up multiple times
+            Thread currentThread = Thread.CurrentThread;
+            if (executionContext != null && !executionContext.m_isDefault)
```
This should never be false; so could just change to Assert

kouvel replied:
If changed to assert, it could probably use a name change as well to reflect that a default context cannot be passed in

benaadams commented on src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:
```diff
@@ -282,6 +292,139 @@ internal static void RunInternal<TState>(ExecutionContext executionContext, Cont
+
+            // For QUWICallbacks there are no extra steps and ThreadPoolWorkQueue.Dispatch 
+            // will reset EC and SyncCtx back to defaults, so it doesn't need to be done here.
+        }
```
Realised can slim further

benaadams commented on src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:
```diff
@@ -566,6 +570,17 @@ internal static bool Dispatch()
                     workQueue.EnsureThreadRequested();
 
+                    // Start on clean ExecutionContext and SynchronizationContext
+                    if (currentThread.ExecutionContext != null)
```
No write barriers for setting things to explicitly to `null`, avoid check

benaadams commented:
`ExecutionContext:RunFromThreadPoolDispatchLoop` (230 bytes asm)
`ExecutionContext:RunForThreadPoolUserWorkItem` (108 bytes asm)
`ExecutionContext:RunForThreadPoolUserWorkItemWithDefaultContext` (71 bytes asm)

With last one's asm being:
```asm
G_M18711_IG01:
       4883EC28             sub      rsp, 40
       90                   nop      
       4889542438           mov      gword ptr [rsp+38H], rdx

G_M18711_IG02:
       498B10               mov      rdx, gword ptr [r8]
       488B442438           mov      rax, gword ptr [rsp+38H]
       488B4808             mov      rcx, gword ptr [rax+8]
       FF5018               call     qword ptr [rax+24]Action`1:Invoke(ref):this
       E800000000           call     Thread:GetCurrentThreadNative():ref
       488B4808             mov      rcx, gword ptr [rax+8]
       4885C9               test     rcx, rcx
       741B                 je       SHORT G_M18711_IG04
       4883791000           cmp      gword ptr [rcx+16], 0
       7414                 je       SHORT G_M18711_IG04
       33D2                 xor      rdx, rdx
       48895008             mov      gword ptr [rax+8], rdx
       488D0500000000       lea      rax, [(reloc)]

G_M18711_IG03:
       4883C428             add      rsp, 40
       48FFE0               rex.jmp  rax           ; tailcall ExecutionContext:OnValuesChanged

G_M18711_IG04:
       4883C428             add      rsp, 40
       C3                   ret    
```

benaadams commented:
> Where are we on determining perf impact (positive or negative)? 😄

Working on it 😄

benaadams commented:
Need to come up with a more deterministic non-random, yet non-colliding way to look at local queues for stealing 😉 

Testing 2,621,440 calls, with GCs after 262,144 calls.
Operations per second on 8 Cores

**Default Context**
```
                                                                             Parallelism
                                  Serial          2x         16x         64x        512x
Pre
                                  
QUWI                             4.191 M     4.180 M     4.282 M     4.317 M     6.482 M
- Depth    2                     6.825 M     6.838 M     6.462 M     5.694 M     5.642 M
- Depth   16                     7.095 M     5.388 M     6.119 M     6.297 M     6.130 M
- Depth   64                     6.009 M     5.873 M     6.256 M     7.074 M     6.547 M
- Depth  512                     7.109 M     7.342 M     7.683 M     6.498 M     7.480 M

Post

QUWI                             2.339 M     6.054 M     5.505 M     5.059 M     6.470 M
- Depth    2                     8.208 M     7.249 M     5.050 M     5.054 M     6.956 M
- Depth   16                     1.338 M     1.836 M     3.396 M     4.715 M     7.184 M
- Depth   64                     2.822 M     4.228 M     5.661 M     8.282 M     8.424 M
- Depth  512                     7.145 M     7.244 M     7.062 M     8.976 M     8.131 M

Pre

QUWI Local(False)                6.458 M     7.989 M     4.901 M     7.121 M    10.382 M
- Depth    2                     8.038 M     6.109 M     3.829 M     4.121 M     6.288 M
- Depth   16                     1.542 M     2.155 M     3.808 M     4.967 M     5.962 M
- Depth   64                     3.637 M     4.107 M     6.095 M     6.190 M     6.970 M
- Depth  512                     6.959 M     7.085 M     7.186 M     7.781 M     6.161 M

Post

QUWI Local(False)                3.923 M     7.590 M     5.269 M     6.512 M     8.383 M
- Depth    2                     8.308 M     6.950 M     4.354 M     5.873 M     6.583 M
- Depth   16                     1.209 M     3.833 M     7.157 M     6.623 M     6.756 M
- Depth   64                     6.034 M     7.219 M     6.871 M     6.870 M     7.576 M
- Depth  512                     7.426 M     8.139 M     7.768 M     8.282 M     7.499 M

Pre
 
QUWI Local(True)                 3.043 M     8.734 M    32.179 M    32.519 M    32.030 M
- Depth    2                     7.911 M    14.130 M    31.797 M    31.808 M    31.197 M
- Depth   16                    26.208 M    28.784 M    32.980 M    32.872 M    31.787 M
- Depth   64                    30.707 M    31.577 M    33.126 M    32.949 M    32.120 M
- Depth  512                    31.114 M    32.359 M    32.492 M    32.633 M    31.507 M
 
Post

QUWI Local(True)                 2.558 M     7.064 M    35.461 M    35.784 M    35.280 M
- Depth    2                     7.878 M    14.664 M    33.128 M    33.754 M    33.489 M
- Depth   16                    27.958 M    30.948 M    35.433 M    35.381 M    34.430 M
- Depth   64                    33.232 M    34.118 M    35.644 M    35.498 M    34.751 M
- Depth  512                    34.629 M    34.881 M    35.303 M    35.105 M    34.229 M

Pre
   
Task.Run Chain                 606.116 k     1.010 M     4.385 M     4.612 M     4.745 M
- Depth    2                   611.915 k     1.069 M     4.352 M     4.724 M     4.721 M
- Depth   16                   588.295 k     1.260 M     4.614 M     4.674 M     4.879 M
- Depth   64                   403.546 k     1.171 M     4.780 M     4.690 M     4.778 M
- Depth  512                   666.265 k     1.445 M     3.556 M     3.659 M     3.679 M

Post

Task.Run Chain                 610.123 k     1.058 M     5.080 M     5.078 M     5.091 M
- Depth    2                   614.935 k     1.119 M     4.987 M     4.888 M     4.995 M
- Depth   16                   610.214 k     1.344 M     4.664 M     4.886 M     5.052 M
- Depth   64                   406.340 k     1.273 M     4.714 M     4.913 M     4.899 M
- Depth  512                   692.124 k     1.517 M     3.667 M     3.593 M     3.659 M

Pre

Task.Run Fanout (WhenAll)      312.163 k   572.976 k     2.286 M     2.412 M     2.442 M
- Depth    2                   480.911 k     1.042 M     2.939 M     2.950 M     2.970 M
- Depth   16                     1.396 M     2.639 M     3.832 M     3.745 M     3.777 M
- Depth   64                     1.846 M     2.973 M     3.870 M     3.807 M     3.750 M
- Depth  512                     2.021 M     3.127 M     3.844 M     3.897 M     3.627 M

Post

Task.Yield Chain               776.884 k     1.495 M     6.719 M     6.868 M     7.231 M
- Depth    2                     1.041 M     2.104 M     6.742 M     7.084 M     7.404 M
- Depth   16                     1.494 M     3.909 M     7.447 M     7.878 M     7.627 M
- Depth   64                     1.504 M     4.821 M     6.438 M     5.806 M     6.170 M
- Depth  512                     1.404 M     5.078 M     4.634 M     4.978 M     4.410 M
```

**Async Locals**
```
                                                                             Parallelism
                                  Serial          2x         16x         64x        512x
Pre
                                  
QUWI                             2.159 M     4.643 M     5.276 M     4.226 M     4.800 M
- Depth    2                     6.932 M     6.008 M     6.024 M     5.315 M     7.067 M
- Depth   16                     5.584 M     5.655 M     5.341 M     6.152 M     6.922 M
- Depth   64                     6.266 M     6.513 M     6.370 M     6.727 M     6.177 M
- Depth  512                     6.423 M     6.196 M     5.439 M     6.503 M     5.927 M

Post

QUWI                             9.910 M     7.060 M     5.631 M     5.394 M     9.946 M
- Depth    2                     7.636 M     6.716 M     4.538 M     4.767 M     6.661 M
- Depth   16                     1.178 M     2.554 M     3.767 M     4.972 M     5.469 M
- Depth   64                     2.922 M     3.441 M     5.803 M     7.112 M     7.604 M
- Depth  512                     6.876 M     6.943 M     7.670 M     8.368 M     7.421 M

Pre

QUWI Local(False)                2.132 M     5.175 M     5.642 M     4.660 M     5.505 M
- Depth    2                     8.154 M     7.214 M     4.906 M     4.616 M     6.091 M
- Depth   16                     1.243 M     2.329 M     4.603 M     5.024 M     6.765 M
- Depth   64                     3.906 M     3.927 M     6.456 M     6.595 M     6.436 M
- Depth  512                     6.685 M     7.950 M     7.299 M     7.591 M     6.848 M

Post

QUWI Local(False)                9.913 M    10.122 M     6.188 M     5.765 M    10.431 M
- Depth    2                     8.058 M     7.948 M     4.671 M     4.783 M     6.028 M
- Depth   16                     1.430 M     2.354 M     4.219 M     4.444 M     6.836 M
- Depth   64                     3.375 M     4.103 M     6.063 M     6.812 M     7.468 M
- Depth  512                     7.479 M     6.708 M     7.245 M     6.963 M     7.484 M

Pre
 
QUWI Local(True)                 2.381 M     8.435 M    27.338 M    27.622 M    27.222 M
- Depth    2                     7.369 M    12.570 M    26.683 M    26.937 M    26.132 M
- Depth   16                    22.508 M    23.829 M    27.681 M    27.764 M    26.522 M
- Depth   64                    25.694 M    25.743 M    28.312 M    27.940 M    26.301 M
- Depth  512                    26.232 M    26.463 M    27.275 M    26.924 M    26.324 M
 
Post

QUWI Local(True)                 2.587 M     8.534 M    31.908 M    32.157 M    31.246 M
- Depth    2                     7.589 M    13.882 M    30.477 M    30.170 M    30.225 M
- Depth   16                    25.248 M    27.217 M    31.293 M    31.378 M    30.412 M
- Depth   64                    30.127 M    30.507 M    31.188 M    30.909 M    30.313 M
- Depth  512                    29.892 M    30.400 M    30.437 M    29.000 M    29.247 M

Pre
   
Task.Run Chain                 556.239 k   857.772 k     3.118 M     3.755 M     3.901 M
- Depth    2                   542.911 k   943.881 k     3.306 M     3.811 M     4.012 M
- Depth   16                   526.391 k     1.161 M     4.023 M     4.203 M     4.306 M
- Depth   64                   377.318 k     1.090 M     3.585 M     4.209 M     4.191 M
- Depth  512                   616.084 k     1.336 M     2.929 M     3.066 M     3.161 M

Post

Task.Run Chain                 570.187 k   875.595 k     3.418 M     3.832 M     3.973 M
- Depth    2                   555.599 k   970.812 k     3.621 M     3.782 M     4.039 M
- Depth   16                   546.947 k     1.166 M     3.804 M     4.013 M     4.169 M
- Depth   64                   375.050 k     1.064 M     3.836 M     3.992 M     3.938 M
- Depth  512                   628.206 k     1.347 M     2.861 M     3.092 M     3.088 M

Pre

Task.Run Fanout (WhenAll)      305.516 k   544.934 k     2.145 M     2.380 M     2.350 M
- Depth    2                   468.987 k   985.685 k     2.510 M     2.770 M     2.794 M
- Depth   16                     1.266 M     2.412 M     3.418 M     3.336 M     3.262 M
- Depth   64                     1.626 M     2.724 M     3.487 M     3.461 M     3.285 M
- Depth  512                     1.726 M     2.865 M     3.397 M     3.388 M     3.345 M

Post

Task.Yield Chain               800.458 k     1.425 M     6.774 M     6.569 M     6.915 M
- Depth    2                     1.023 M     2.130 M     5.928 M     7.320 M     7.240 M
- Depth   16                     1.490 M     4.028 M     7.645 M     7.687 M     6.601 M
- Depth   64                     1.468 M     4.986 M     6.351 M     5.387 M     5.623 M
- Depth  512                     1.429 M     4.980 M     5.426 M     5.031 M     4.340 M
```

Variance aside, looks better overall. 

Though my big takeaway is favour the new generic overload (with `preferLocal: false` to achieve same behaviour)
`QueueUserWorkItem<TState>(Action<TState> callBack, TState state, preferLocal: false)` over `QueueUserWorkItem(WaitCallback callBack, object state)`

kouvel commented on src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:
```diff
@@ -333,17 +329,14 @@ internal static void RunFromThreadPoolDispatchLoop(ExecutionContext executionCon
-            }
+
+            // Restore changed SynchronizationContext back to Default
+            currentThread1.SynchronizationContext = null;
```
Isn't it better to check for null before writing to it like you had before?

benaadams replied:
Since its `null` constant the write barrier doesn't get involved; so its whether `read+cmp+je` is better than a single non-dependent write (e.g.)
```asm
       49837F0800           cmp      gword ptr [r15+8], 0  ; is null check
       7406                 je       SHORT G_M60392_IG10 ; jump taken
       33C9                 xor      rcx, rcx
       49894F08             mov      gword ptr [r15+8], rcx  ; set null
G_M60392_IG10:
       49837F1000           cmp      gword ptr [r15+16], 0  ; is null check
       7406                 je       SHORT G_M60392_IG11 ; jump taken
       33C9                 xor      rcx, rcx
       49894F10             mov      gword ptr [r15+16], rcx  ; set null
``` 
vs
```asm
       33C9                 xor      rcx, rcx
       49894F08             mov      gword ptr [r15+8], rcx  ; set null
       49894F10             mov      gword ptr [r15+16], rcx  ; set null
```

kouvel replied:
Ah ok, makes sense

AndyAyersMS replied:
I'd say the second is quite likely better. 

Stores are generally cheap and branches and loads can be expensive.

kouvel commented on src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:
```diff
@@ -356,73 +349,60 @@ internal static void RunFromThreadPoolDispatchLoop(ExecutionContext executionCon
 
             ExecutionContext currentExecutionCtx = currentThread.ExecutionContext;
-            if (currentExecutionCtx != null)
+            if (currentExecutionCtx.HasChangeNotifications)
```
Couldn't the callback have changed the execution context back to default (null)?

benaadams replied:
No, as the only way to get back to the default context is being returned via the `ExecutionContext.Run{Variants}`; setting the last `AsyncLocal<T>` to `null` will leave you on a non-default context with an `EmptyAsyncLocalValueMap` as the locals

kouvel commented on src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:
```diff
@@ -356,73 +349,60 @@ internal static void RunFromThreadPoolDispatchLoop(ExecutionContext executionCon
             Thread currentThread = Thread.CurrentThread;
             ExecutionContext currentExecutionCtx = currentThread.ExecutionContext;
-            if (currentExecutionCtx != null)
+            if (currentExecutionCtx != null && currentExecutionCtx.HasChangeNotifications)
```
If the current non-default context does not have notifications it would not be reset to null, but unless Dispatch nulls it after this work item completes, would it not be possible for a timer callback to observe a non-null context on start?

kouvel replied:
Which I guess is ok because unmanaged work items don't reset either, nevermind. In that case though, why reset the thread's execution context to null at all in `RunFromThreadPoolDispatchLoop` and `RunForThreadPoolUserWorkItem`?

benaadams replied:
For both they set back to the original prior to the change notifications running; and for the `RunFromThreadPoolDispatchLoop` since the more async machinery runs after; it has to be set back before it returns.

There is even an AsyncLocal test that mutates the EC in the change handler (but changing an asynclocal) which it then checks leaks out of the await.

kouvel replied:
The differences seem to be subtle, I'm not sure it's worth distinguishing those. Why not keep it simple and always reset the execution context? It'll be an extra write but it'll keep the methods symmetric and it wouldn't require extra care to use them. If we really want to optimize out the write I would suggest the opposite, and not reset the execution context in any of these methods, and leave that to the caller for all cases (with a comment).

kouvel replied:
Resetting before notifications I guess is necessary as it is the previous behavior, but when notifications are not necessary

benaadams replied:
Yeah, so it doesn't reset unless there are notifications; because the current context can be observed in the notification handler. Except in the non UserWorkItem case where it always resets because they don't return directly to the Dispatch loop (e.g. Task) so again the context will be observed (and passed out of any awaits)

benaadams replied:
Done

kouvel commented:
It looks like `QueueUserWorkItemCallback.Execute` and `QueueUserWorkItemCallback<TState>.Execute` call the callback directly without going through `ExecutionContext`, so is it possible that context changes during those callbacks would not get change notifications? Or should those cases be using the `WithDefaultContext` version of those types?

benaadams commented:
> It looks like QueueUserWorkItemCallback.Execute and QueueUserWorkItemCallback<TState>.Execute call the callback directly without going through ExecutionContext

Only when the EC is `null` which indicates they are flow suppressed otherwise they go via `ExecutionContext.RunForThreadPoolUserWorkItem`

>  is it possible that context changes after those callbacks complete would not get change notifications? 

Yes; though currently you'd mutate the ThreadPool thread's context https://github.com/dotnet/corefx/issues/32695 and receive 2 sets of change notification for every item that runs on the ThreadPool forever more (for example it would be moving from mutated context to Default; run item; Default to mutated context). Worse still if the change notification changed an async local; it would then inject it into every context that ran on that ThreadPool thread spreading like poison until it was in every context.

kouvel commented:
> Only when the EC is null which indicates they are flow suppressed otherwise they go via ExecutionContext.RunForThreadPoolUserWorkItem

I guess flow-suppressed is a one-way thing currently, if the callback adds an `AsyncLocal` it wouldn't get change notifications for when the reset happens from this change. Should it?

> Yes; though currently you'd mutate the ThreadPool thread's context dotnet/corefx#32695 and receive 2 sets of change notification for every item that runs on the ThreadPool forever more (for example it would be moving from mutated context to Default; run item; Default to mutated context).

Doesn't this change fix that by resetting the context between work items?

kouvel commented:
> I guess flow-suppressed is a one-way thing currently, if the callback adds an AsyncLocal it wouldn't get change notifications for when the reset happens from this change. Should it?

On non-threadpool threads it may be breaking but on thread pool threads for QUWI work items it may make sense to send notifications since we'd reset to null between work items with this change anyway

benaadams commented:
> > I guess flow-suppressed is a one-way thing currently, if the callback adds an AsyncLocal it wouldn't get change notifications for when the reset happens from this change. Should it?

> On non-threadpool threads it may be breaking but on thread pool threads for QUWI work items it may make sense to send notifications since we'd reset to null between work items with this change anyway

On a non-threadpool thread it won't send an `contextChanged: true` change notification for flow suppressed items as the thread will just exit (other then when `ExecutionContext.Run` is manually used on that thread)

So this would just be replicating that behaviour on the threadpool thread? The flow supressed item would behave as if it was on its own dedicated thread (though it would be threadpool). Again any manual use of  `ExecutionContext.Run` as part of the workitem run will still fire.

The alternative would be to reset the context at the end of the Dispatch loop; fire any notifications, then reset again (though it wouldn't then fire for changes made in the notification handler, unless the reset was in a loop, but then it could be an infinite loop)

kouvel commented:
> On a non-threadpool thread it won't send an `contextChanged: true` change notification for flow suppressed items as the thread will just exit (other then when `ExecutionContext.Run` is manually used on that thread)

> So this would just be replicating that behaviour on the threadpool thread? The flow supressed item would behave as if it was on its own dedicated thread (though it would be threadpool). Again any manual use of  `ExecutionContext.Run` as part of the workitem run will still fire.

In non-threadpool cases the thread may not exit. It was consistent to the point that if change notifications are not sent, then the EC is still in effect for whatever work happens on that thread later. This change is introducing a reset, which breaks the above assumption in flow-suppressed cases that would not get change notifications. Not sure if that would be ok.

I'm not sure about changes that occur during change notifications. I don't think anything special has to be done for that corner case for now, as it is already problematic and unexpected.

benaadams commented:
An alternative could be to run flow suppressed items in a Default context, rather than directly?

benaadams commented:
An alternative could be to run flow suppressed items in a Default context, rather than directly? Then the notifications would fire as it exited?

kouvel commented:
> https://github.com/dotnet/corefx/issues/25477#issuecomment-346866897
> And everything else after here makes no sense, since the only place that set a value has finished, and already set the value back to null;

I think that issue is referring to the problem in https://github.com/dotnet/corefx/issues/30275, which has been fixed.

> An alternative could be to run flow suppressed items in a Default context, rather than directly? Then the notifications would fire as it exited?

Yes that's what I was thinking, though it would be unfortunate. @stephentoub, thoughts?

benaadams commented:
> I think that issue is referring to the problem in dotnet/corefx#30275, which has been fixed.

Yeah, but the same thing happens, e.g.

```csharp
class Program
{
    static void Main(string[] args)
    {
        SemaphoreSlim semaphore = new SemaphoreSlim(0);

        ThreadPool.UnsafeQueueUserWorkItem((s) => 
        {
            var local = new AsyncLocal<int>(ValueChangedHandler) { Value = 1 };
            semaphore.Release();
        }, null);

        semaphore.Wait();
        for (var i = 0; i < 100; i++)
        {
            ThreadPool.QueueUserWorkItem((s) => { semaphore.Release(); }, null);
            semaphore.Wait();
        }
    }

    private static void ValueChangedHandler(AsyncLocalValueChangedArgs<int> obj)
    {
        Console.WriteLine($"Prev: {obj.PreviousValue}, Current: {obj.CurrentValue}, ContextChange: {obj.ThreadContextChanged}");
    }
}
```
Produces this output which might be unexpected
```
Prev: 0, Current: 1, ContextChange: False
Prev: 1, Current: 0, ContextChange: True
Prev: 0, Current: 1, ContextChange: True
Prev: 1, Current: 0, ContextChange: True
Prev: 0, Current: 1, ContextChange: True
Prev: 1, Current: 0, ContextChange: True
Prev: 0, Current: 1, ContextChange: True
Prev: 1, Current: 0, ContextChange: True
Prev: 0, Current: 1, ContextChange: True
Prev: 1, Current: 0, ContextChange: True
Prev: 0, Current: 1, ContextChange: True
```

kouvel commented on src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:
```diff
@@ -605,7 +613,11 @@ internal static bool Dispatch()
+
+                    // Release refs and return to clean ExecutionContext and SynchronizationContext
+                    outerWorkItem = workItem = null;
+                    currentThread.ExecutionContext = null;
```
> An alternative could be to run flow suppressed items in a Default context, rather than directly? Then the notifications would fire as it exited?

I think more generally, any change to execution context should coincide with sending notifications. Here too it should be checking for and sending notifications if necessary.

kouvel replied:
And then the Dispatch-called methods won't have to do anything after invoking the work item.

kouvel replied:
> The alternative would be to reset the context at the end of the Dispatch loop; fire any notifications, then reset again (though it wouldn't then fire for changes made in the notification handler, unless the reset was in a loop, but then it could be an infinite loop)

As you mentioned, since we're wanting to guarantee that thread pool threads are restored to a null EC, after sending change notifications, we could reset the EC again without sending change notifications. Changing the EC in a changed handler would be polluting anyway and in hindsight it may have been better generally to change the EC after sending change notifications, but I guess that would be more breaking.

benaadams replied:
k will make that change

kouvel commented on src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:
```diff
@@ -388,16 +380,23 @@ internal static void RunForThreadPoolUserWorkItemWithDefaultContext<TState>(Acti
             }
 
-            // ThreadPoolWorkQueue.Dispatch will reset EC and SyncCtx back to default
+            // Reset back to defaults in case the Change handler changed the contexts
```
This is the main reset, maybe the comment could be tweaked or it could be written a bit differently:

```c#
// Reset to defaults
currentThread.ExecutionContext = null;
currentThread.SynchronizationContext = null;

if (currentExecutionCtx != null && currentExecutionCtx.HasChangeNotifications)
{
    OnValuesChanged(currentExecutionCtx, nextExecutionCtx: null);

    // Reset to defaults again without change notifications in case the Change handler changed the contexts
    currentThread.ExecutionContext = null;
    currentThread.SynchronizationContext = null;
}
```

benaadams replied:
I like that

benaadams replied:
At this point default context UWI can just run rather than using EC.Run, which makes it the same as unsafe run, so unsafe can use the `QueueUserWorkItemCallbackDefaultContext` version?

benaadams replied:
Since its not really doing anything
```csharp
internal static void RunForThreadPoolUserWorkItemWithDefaultContext<TState>(Action<TState> callback, in TState state)
{
    CheckThreadPoolAndContextsAreDefault();

    callback.Invoke(state);

    // ThreadPoolWorkQueue.Dispatch will handle notifications and reset EC and SyncCtx back to default
}
```

kouvel replied:
Yea true any of the default EC cases could use `QueueUserWorkItemCallbackDefaultContext` instead

benaadams replied:
So would be the last change now? https://github.com/dotnet/coreclr/pull/20308/commits/b412b4f453a05f7e21ef29a9100446ce743705cc

benaadams commented:
@dotnet-bot test

benaadams commented:
@dotnet-bot test OSX10.12 x64 Checked Innerloop Build and Test

benaadams commented:
Added an aggressive inline to 
```
[FAILED: unprofitable inline] ExecutionContext:ResetThreadPoolThread(ref)
```
since its always called and fairly short

benaadams commented:
Infra issue

@dotnet-bot test Ubuntu16.04 arm64 Cross Checked no_tiered_compilation_innerloop Build and Test

benaadams commented:
@dotnet-bot test Ubuntu16.04 arm64 Cross Checked no_tiered_compilation_innerloop Build and Test

benaadams commented:
@dotnet-bot test Ubuntu16.04 arm64 Cross Checked no_tiered_compilation_innerloop Build and Test

benaadams commented:
Can't seem to retrigger that test

benaadams commented:
Additional changes:

`Task.ExecuteFromThreadPool()` is an internal api; since we have the current thread in the Dispatch loop now, can change it to `Task.ExecuteFromThreadPool(Thread threadPoolThread)` and pass in the thread to give to the `ExecutionContext` and skip the look up of `Thread.CurrentThread` there.

Also `AwaitTaskContinuation` can run the `m_action` directly for Default context in addition to flow suppressed context; like `QueueUserWorkItemCallbackDefaultContext` as it doesn't execute anything after; now the Dispatch loop handles EC restore.

kouvel commented on src/System.Private.CoreLib/src/System/Threading/ThreadPool.cs:
```diff
@@ -1026,25 +1008,24 @@ internal sealed class QueueUserWorkItemCallbackDefaultContext : QueueUserWorkIte
             base.Execute();
-            ExecutionContext.RunForThreadPoolUserWorkItemWithDefaultContext(s_executionContextShim, this);
+
+            WaitCallback callback = _callback;
```
Since this expects to run in the default context it would be good to assert that the current contexts are default (the `CheckThreadPoolAndContextsAreDefault` that was there before)

benaadams replied:
Added (x3)

kouvel commented on src/System.Private.CoreLib/shared/System/Threading/ExecutionContext.cs:
```diff
@@ -293,12 +293,11 @@ internal static void RunInternal<TState>(ExecutionContext executionContext, Cont
         }
 
-        internal static void RunFromThreadPoolDispatchLoop(ExecutionContext executionContext, ContextCallback callback, object state)
+        internal static void RunFromThreadPoolDispatchLoop(Thread threadPoolThread, ExecutionContext executionContext, ContextCallback callback, object state)
```
Could you please add an assert that the thread passed in is the same as the current thread?

kouvel replied:
And similarly elsewhere

kouvel commented on src/System.Private.CoreLib/src/System/Runtime/CompilerServices/AsyncMethodBuilder.cs:
```diff
@@ -562,13 +562,13 @@ private void MoveNext(bool isThreadPool)
                 else
                 {
-                    if (isThreadPool)
+                    if (threadPoolThread is null)
```
Is `is null` better than `== null`? That would seem odd to me. (similarly elsewhere)

benaadams replied:
C# 7 ism; `== null` will call the Equals overload if its defined, `is null` always does `ReferenceEquals(...,, null)`

kouvel replied:
I see

benaadams commented:
```diff
 Testing 2,621,440 calls, with GCs after 262,144 calls.
 Operations per second on 8 Cores
                                                                              Parallelism
                                   Serial          2x         16x         64x        512x
 Default Context
 
-QUWI                             6.681 M     8.555 M     3.985 M     4.758 M     5.782 M
+QUWI                             3.671 M     8.608 M     4.220 M     5.211 M     7.647 M
-- Depth    2                     7.998 M     5.729 M     4.925 M     6.618 M     6.060 M
+- Depth    2                     7.550 M     5.887 M     4.488 M     6.327 M     6.486 M
-- Depth   16                     6.192 M     7.280 M     4.860 M     6.285 M     7.200 M
+- Depth   16                     6.107 M     6.047 M     6.136 M     5.593 M     7.320 M
-- Depth   64                     6.181 M     6.839 M     7.441 M     7.694 M     6.646 M
+- Depth   64                     6.838 M     5.551 M     7.790 M     6.847 M     6.762 M
-- Depth  512                     8.059 M     7.184 M     5.862 M     8.137 M     6.518 M
+- Depth  512                     6.627 M     7.838 M     7.111 M     6.825 M     7.208 M

-QUWI Local(False)                3.079 M     4.101 M     4.765 M     4.555 M     7.063 M
+QUWI Local(False)                4.007 M     5.484 M     5.068 M     5.026 M     7.953 M
-- Depth    2                     7.403 M     5.829 M     6.403 M     5.984 M     6.438 M
+- Depth    2                     7.729 M     5.928 M     6.082 M     5.647 M     6.536 M
-- Depth   16                     5.784 M     6.940 M     6.330 M     5.867 M     5.511 M
+- Depth   16                     6.337 M     5.279 M     6.098 M     6.354 M     6.785 M
-- Depth   64                     6.230 M     6.796 M     6.533 M     8.058 M     5.882 M
+- Depth   64                     6.559 M     5.957 M     7.666 M     6.681 M     8.449 M
-- Depth  512                     6.541 M     7.342 M     6.490 M     6.029 M     7.363 M
+- Depth  512                     8.242 M     7.823 M     6.137 M     6.819 M     5.932 M

-QUWI Local(True)                 2.694 M     6.481 M    28.697 M    29.211 M    28.719 M
+QUWI Local(True)                 3.098 M     7.934 M    36.140 M    36.854 M    36.004 M
-- Depth    2                     7.539 M    13.679 M    28.713 M    28.225 M    27.684 M
+- Depth    2                     7.875 M    14.842 M    35.850 M    35.794 M    34.711 M
-- Depth   16                    24.621 M    26.177 M    29.206 M    29.559 M    28.799 M
+- Depth   16                    29.033 M    31.585 M    37.443 M    37.150 M    35.621 M
-- Depth   64                    28.158 M    28.537 M    29.804 M    29.546 M    29.154 M
+- Depth   64                    34.548 M    34.971 M    37.523 M    37.465 M    36.411 M
-- Depth  512                    28.861 M    28.969 M    29.670 M    29.113 M    28.429 M
+- Depth  512                    36.064 M    36.705 M    36.913 M    36.792 M    35.728 M

-Task.Run Chain                 559.677 k   916.428 k     3.703 M     4.065 M     4.372 M
+Task.Run Chain                 604.549 k     1.062 M     4.512 M     4.722 M     4.939 M
-- Depth    2                   498.474 k     1.000 M     3.871 M     4.316 M     4.474 M
+- Depth    2                   607.223 k     1.121 M     5.044 M     5.095 M     5.132 M
-- Depth   16                   525.809 k     1.196 M     3.569 M     4.424 M     4.623 M
+- Depth   16                   639.864 k     1.404 M     4.929 M     5.365 M     5.373 M
-- Depth   64                   387.807 k     1.065 M     4.335 M     4.346 M     4.414 M
+- Depth   64                   417.029 k     1.335 M     5.190 M     5.240 M     5.362 M
-- Depth  512                   634.539 k     1.433 M     3.365 M     3.730 M     3.741 M
+- Depth  512                   707.908 k     1.566 M     3.612 M     3.894 M     4.071 M

-Task.Run Fanout (WhenAll)      305.513 k   558.586 k     2.369 M     2.314 M     2.411 M
+Task.Run Fanout (WhenAll)      337.125 k   615.334 k     2.567 M     2.506 M     2.551 M
-- Depth    2                   471.856 k     1.029 M     2.854 M     2.820 M     2.861 M
+- Depth    2                   518.558 k     1.106 M     3.004 M     3.058 M     3.058 M
-- Depth   16                     1.366 M     2.563 M     3.640 M     3.605 M     3.616 M
+- Depth   16                     1.476 M     2.747 M     3.862 M     3.857 M     3.882 M
-- Depth   64                     1.808 M     2.892 M     3.692 M     3.740 M     3.596 M
+- Depth   64                     1.934 M     3.085 M     4.041 M     3.977 M     3.801 M
-- Depth  512                     2.032 M     3.043 M     3.751 M     3.676 M     3.574 M
+- Depth  512                     2.141 M     3.216 M     3.972 M     3.972 M     3.687 M

-Task.Yield Chain               767.049 k     1.410 M     6.642 M     6.647 M     6.902 M
+Task.Yield Chain               828.771 k     1.543 M     6.547 M     7.425 M     7.448 M
-- Depth    2                     1.004 M     2.087 M     7.326 M     7.374 M     7.388 M
+- Depth    2                     1.092 M     2.196 M     7.445 M     7.453 M     7.433 M
-- Depth   16                     1.563 M     4.095 M     7.484 M     7.796 M     6.469 M
+- Depth   16                     1.583 M     4.096 M     7.678 M     8.080 M     7.955 M
-- Depth   64                     1.497 M     5.069 M     5.225 M     6.085 M     4.869 M
+- Depth   64                     1.542 M     4.991 M     6.320 M     6.235 M     4.833 M
-- Depth  512                     1.439 M     5.250 M     4.178 M     4.629 M     4.272 M
+- Depth  512                     1.552 M     5.363 M     5.730 M     5.894 M     4.929 M

 Async Locals
 
-QUWI                             7.651 M     7.490 M     7.211 M     5.130 M     6.934 M
+QUWI                             4.126 M     6.018 M     6.665 M     5.047 M     6.866 M
-- Depth    2                     7.738 M     6.146 M     4.341 M     5.392 M     5.433 M
+- Depth    2                     6.730 M     5.878 M     4.537 M     7.217 M     8.080 M
-- Depth   16                   974.033 k     1.232 M     2.820 M     4.567 M     6.110 M
+- Depth   16                     5.873 M     5.414 M     6.573 M     6.047 M     6.225 M
-- Depth   64                     2.192 M     3.127 M     4.207 M     5.142 M     6.095 M
+- Depth   64                     6.381 M     6.440 M     5.795 M     6.692 M     6.384 M
-- Depth  512                     5.566 M     7.265 M     7.197 M     6.998 M     6.080 M
+- Depth  512                     6.532 M     7.274 M     6.194 M     7.267 M     6.658 M

-QUWI Local(False)                7.931 M     6.849 M     7.185 M     5.626 M     6.086 M
+QUWI Local(False)                3.251 M     5.675 M     6.785 M     4.560 M     6.526 M
-- Depth    2                     7.928 M     7.140 M     4.934 M     4.430 M     4.823 M
+- Depth    2                     8.101 M     7.173 M     5.492 M     5.050 M     6.177 M
-- Depth   16                     1.095 M     1.523 M     3.367 M     4.353 M     6.872 M
+- Depth   16                     6.134 M     5.794 M     6.548 M     6.536 M     5.955 M
-- Depth   64                     2.354 M     4.319 M     5.253 M     5.255 M     7.181 M
+- Depth   64                     6.018 M     6.500 M     6.248 M     5.816 M     7.198 M
-- Depth  512                     6.328 M     6.416 M     7.896 M     8.289 M     7.300 M
+- Depth  512                     6.361 M     6.692 M     7.258 M     6.995 M     6.621 M

-QUWI Local(True)                 2.696 M     8.880 M    29.852 M    30.371 M    29.782 M
+QUWI Local(True)                 2.509 M     8.550 M    29.825 M    30.928 M    30.726 M
-- Depth    2                     7.883 M    14.216 M    28.745 M    28.718 M    27.964 M
+- Depth    2                     7.529 M    13.840 M    30.561 M    30.160 M    30.084 M
-- Depth   16                    24.383 M    26.722 M    29.689 M    29.737 M    29.164 M
+- Depth   16                    25.469 M    27.823 M    31.548 M    31.715 M    31.091 M
-- Depth   64                    27.669 M    28.243 M    29.964 M    29.286 M    28.766 M
+- Depth   64                    29.957 M    30.074 M    31.932 M    31.740 M    30.601 M
-- Depth  512                    28.094 M    28.727 M    29.069 M    29.152 M    28.468 M
+- Depth  512                    30.588 M    30.483 M    31.391 M    31.546 M    30.468 M

-Task.Run Chain                 540.019 k   896.002 k     3.720 M     4.060 M     4.096 M
+Task.Run Chain                 573.983 k   892.269 k     3.614 M     3.779 M     3.914 M
-- Depth    2                   536.054 k   948.196 k     3.497 M     3.851 M     4.093 M
+- Depth    2                   558.895 k   977.010 k     3.343 M     3.870 M     4.104 M
-- Depth   16                   515.707 k     1.139 M     3.670 M     4.171 M     4.296 M
+- Depth   16                   563.103 k     1.219 M     4.124 M     4.222 M     4.340 M
-- Depth   64                   379.812 k     1.023 M     3.849 M     4.025 M     4.239 M
+- Depth   64                   391.039 k     1.154 M     4.034 M     4.060 M     4.242 M
-- Depth  512                   599.065 k     1.357 M     2.912 M     3.036 M     3.186 M
+- Depth  512                   664.986 k     1.391 M     2.835 M     2.950 M     3.273 M

-Task.Run Fanout (WhenAll)      296.167 k   542.683 k     1.994 M     2.286 M     2.291 M
+Task.Run Fanout (WhenAll)      308.828 k   555.943 k     2.261 M     2.278 M     2.311 M
-- Depth    2                   435.371 k   986.554 k     2.677 M     2.756 M     2.686 M
+- Depth    2                   461.337 k   993.569 k     2.509 M     2.765 M     2.772 M
-- Depth   16                     1.254 M     2.393 M     3.285 M     3.183 M     3.184 M
+- Depth   16                     1.300 M     2.444 M     3.347 M     3.335 M     3.297 M
-- Depth   64                     1.660 M     2.738 M     3.372 M     3.311 M     3.215 M
+- Depth   64                     1.673 M     2.765 M     3.517 M     3.460 M     3.308 M
-- Depth  512                     1.739 M     2.855 M     3.356 M     3.310 M     3.239 M
+- Depth  512                     1.802 M     2.866 M     3.424 M     3.358 M     3.399 M

-Task.Yield Chain               767.848 k     1.433 M     7.024 M     6.820 M     6.862 M
+Task.Yield Chain               785.584 k     1.481 M     6.394 M     7.127 M     7.126 M
-- Depth    2                   983.623 k     2.016 M     7.051 M     7.386 M     7.358 M
+- Depth    2                     1.048 M     2.084 M     6.214 M     7.172 M     7.352 M
-- Depth   16                     1.547 M     3.901 M     7.444 M     7.596 M     7.357 M
+- Depth   16                     1.527 M     4.001 M     7.137 M     7.635 M     7.630 M
-- Depth   64                     1.506 M     4.868 M     6.010 M     6.190 M     5.460 M
+- Depth   64                     1.549 M     4.846 M     5.735 M     5.735 M     6.009 M
-- Depth  512                     1.494 M     5.138 M     4.998 M     5.696 M     4.370 M
+- Depth  512                     1.495 M     5.187 M     5.126 M     5.855 M     4.257 M
```

benaadams commented:
Updated summary:

Maintain the ThreadPool threads on the Default contexts between work items.

Always restore the Default context on the ThreadPool Dispatch loop after a workitem has run (to clean up any ExecutionContext leakage from changes on flow suppressed workitems, or AsyncLocal change eventhandlers; as well as firing their notifications if they have them)

Store the `CurrentThread` as part of the thread static `ThreadPoolWorkQueueThreadLocals` which are already looked up at the start of the Dispatch loop to avoid an additional lookup via `Thread.CurrentThread`.

As workitems are started on the Default context and are returned to it `QueueUserWorkItemCallbackDefaultContext` items can just be run their callbacks directly rather than via `ExecutionContext.Run` (as the EC don't need to move to Default and is always moved back to Default).

As `QueueUserWorkItemCallbackDefaultContext` now runs items directly; flow suppressed callbacks can use the smaller `QueueUserWorkItemCallbackDefaultContext` rather than `QueueUserWorkItemCallback` with a null context; and handling for flow suppression can be removed from `QueueUserWorkItemCallback`.

As `AwaitTaskContinuation`'s `IThreadPoolWorkItem.Execute` doesn't preform additional work after it completes, it can run `m_action` directly for Default context in addition to the flow suppressed context, rather than going via `ExecutionContext.Run`.

Given that the items on the ThreadPool are always started on the threadpool and restored to it; we can introduce some faster paths than `ExecutionContext:RunInternal` (328 bytes asm).

Introduce `ExecutionContext:RunForThreadPoolUnsafe` (71 bytes asm), for `IThreadPoolWorkItem`s where they need to run on a provided context, but do not need to execute anything after they complete so can rely on the Dispatch loop restore. This includes `QueueUserWorkItemCallback`, `QueueUserWorkItemCallback<TState>` and `AwaitTaskContinuation`.

Introduce `ExecutionContext:RunFromThreadPoolDispatchLoop` (225 bytes asm), for items run from the ThreadPool, so don't need to capture current context (Default) to restore later, however need to do need to restore back to Default after execution as they then perform additional work. This includes 
`Task`/`AsyncStateMachineBox`/`AwaitTaskContinuation`/`Timer`. 

Change `Task.ExecuteFromThreadPool()` to take the thread `Task.ExecuteFromThreadPool(Thread threadPoolThread)` from the ThreadPool Dispatch loop so it can pass it into the `ExecutionContext:RunFromThreadPoolDispatchLoop` overload and avoid the `Thread.CurrentThread` lookup.

benaadams commented:
@kouvel thank you for the reviews and joining me on this journey :)

benaadams commented:
@dotnet-bot test Windows_NT x64 Checked Innerloop Build and Test
